tests/Examples/cleartext/dot_product_8_test.mlir

```diff
@@ -0,0 +1,39 @@
+// RUN: heir-opt %s --mlir-to-secret-arithmetic | FileCheck %s --check-prefix=CHECK_MLIR_TO_SECRET_ARITHMETIC
+
+// RUN: heir-opt %s --mlir-to-secret-arithmetic --tensor-ext-to-tensor --secret-forget-secrets --heir-polynomial-to-llvm \
+// RUN:   | mlir-runner -e test_dot_product -entry-point-result=void \
```

Comment:
Some of the complexity is in the use of mlir-runner. We could instead lower all the way to machine code and use a C++ test harness. See for example https://github.com/google/heir/blob/9c2848bae24aac36dd8d449c1eacfc0dbe965038/tests/Dialect/Arith/Transforms/runner/lower_split.mlir#L2

The bonus of not using mlir-runner is that it alone contributes to more than half the build time of the entire project (because it pulls in all dialects and transforms from upstream). So let's try avoiding to use it.

Reply:
Using c test harness means we have to deal with the memref descriptor for _tensor type_, which is quite implicit.

I had some experience dealing with it when doing the previous version of plaintext execution at https://github.com/ZenithalHourlyRate/heir/commit/617e0d163690710c5ed44840553e9193d936ea33 so I am now able to know which signature I would use, for either raw signature or the `llvm.emit_c_interface` signature, but I would admit that is painful.

Also this introduces clang and mlir-translate dependency, and if we enable it by default (instead of a "manual" tag in BUILD), we will have more build time.

Reply:
I think this is much better, at least in part because `mlir-translate` doesn't have all the LLVM passes in it, and `clang` is a dependency we can get from the host machine if need be, while `mlir-runner` pulls in every dialect and transform from MLIR upstream, plus all the LLVM infrastructure they rely on, which is a LOT.